x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java

```diff
@@ -216,187 +229,175 @@ public WildcardFieldType clone() {
 
-            @Override
-            public boolean equals(Object obj) {
-                PatternStructure other = (PatternStructure) obj;
-                return pattern.equals(other.pattern);
+            return createApproxAndVerifyQuery(approxAutomata, dvAutomaton, wildcardPattern);
+        }
+        
+        @Override
+        public Query regexpQuery(String value, int flags, int maxDeterminizedStates, RewriteMethod method, QueryShardContext context) {
+            if (context.allowExpensiveQueries() == false) {
+                throw new ElasticsearchException(
+                    "[regexp] queries cannot be executed when '" + ALLOW_EXPENSIVE_QUERIES.getKey() + "' is set to false."
+                );
             }
+            failIfNotIndexed();
+            RegExp regex = new RegExp(value, flags);
+            Automaton automaton = regex.toAutomaton(maxDeterminizedStates);
+            ApproximateRegExp ngramRegex = new ApproximateRegExp(toLowerCase(value), flags);
+            Automaton ngramAutomaton = ngramRegex.toAutomaton(maxDeterminizedStates);
 
-
+            return createApproxAndVerifyQuery(ngramAutomaton, automaton, value);
+        }
+        
+        private static String toLowerCase(String string) {
+            return lowercaseNormalizer.normalize(null, string).utf8ToString();
         }
+        
+        
+        Query createApproxAndVerifyQuery(Automaton simplifiedAutomaton, Automaton verifyAutomaton, String expression) {
 
+            Query ngramApproximation = createApproximationQueryFromAutomaton(simplifiedAutomaton);
 
-        @Override
-        public Query wildcardQuery(String wildcardPattern, RewriteMethod method, QueryShardContext context) {
-            PatternStructure patternStructure = new PatternStructure(wildcardPattern);
-            ArrayList<String> tokens = new ArrayList<>();
-
-            for (int i = 0; i < patternStructure.fragments.length; i++) {
-                String fragment = patternStructure.fragments[i];
-                int fLength = fragment.length();
-                if (fLength == 0) {
-                    continue;
+            AutomatonQueryOnBinaryDv verifyingQuery = new AutomatonQueryOnBinaryDv(name(), expression, verifyAutomaton);
+            if (ngramApproximation != null && !(ngramApproximation instanceof MatchAllDocsQuery)) {
+                // We can accelerate execution with the ngram query
+                BooleanQuery.Builder verifyingBuilder = new BooleanQuery.Builder();
+                verifyingBuilder.add(new BooleanClause(ngramApproximation, Occur.MUST));
+                verifyingBuilder.add(new BooleanClause(verifyingQuery, Occur.MUST));
+                return verifyingBuilder.build();
+            }
+            return verifyingQuery;
+        }
+        
+
+        private Query createApproximationQueryFromAutomaton(Automaton simplifiedAutomaton) {
+            FiniteStringsIterator iterator = new FiniteStringsIterator(simplifiedAutomaton);
```

Comment:
I think that's ok for a first iteration but we should look at dividing the boolean query around the automaton's [articulation points](https://en.wikipedia.org/wiki/Biconnected_component). That could optimize a lot the boolean query that we produce on complex automaton.

Reply:
Doesn't that take us back to Nik's code and exploring the transition points in the automaton to understand the graph? I'm not sure how FiniteStringsIterator works in conjunction with that sort of logic.

Reply:
We have an example in Lucene in `QueryBuilder#analyzeGraphBoolean`. I don't think we should do that in the first iteration though so let's assume that we have to visit all the paths for now. 
However I am thinking of one thing that we could try to optimize in this pr, when we have a few transitions that have the same target. For instance `[aA][bB][cD]` that would create 8 paths, we could instead allow each path to contain a few unicode points per positions (2 maybe 3 max) and fork `FiniteStringIterator` to return optimized `IntsRef[][]` ?

Reply:
I'm not sure why we don't normalize case given:

- case is likely to be the biggest cause of logic-branching (based on the existing rules we've seen from the security space) 
- normalising reduces our code complexity and increases the number of expressions we can hope to accelerate
- the ngram search just has to be fast, not 100% correct. Normalising will:
  - dramatically reduce the speed of searches (up to 8 x fewer unique terms)
  - not increase false positives massively for most cases (gut feel on this)

Reply:
Sure but I still think we should consider this change as a follow up. This pr should ensure that we can handle all types of automaton correctly no matter if we lowercase or not. 
Then we can discuss how to handle case insensitive search but that's a different discussions imo.

Reply:
>Then we can discuss how to handle case insensitive search but that's a different discussions imo.

The idea behind normalisation in this context isn't about enabling case insensitivity for end users - it's about optimising the search performance and minimising the complexity of automatons. 
There will inevitably be limits on the number of permutations we can consider and  working with a lower-cased vocabulary will help reduce the numbers.

Reply:
That's assuming that lots of query will contains `[aA][bB]` variations but this could be avoided if we provide a true case-insensitive search. Again, I am not saying we shouldn't do this "optimization" but to me this change would deserve a separate discussion where we'd also see think about exposing the case insensitive search on this field.